virtual_rainforest/core/core_components.py

```diff
@@ -0,0 +1,278 @@
+    configuration. All heights are in metres relative to ground level and the canopy
+    layer heights are defined dynamically by the
+    :class:`~virtual_rainforest.models.plants.PlantsModel`.
+
+    .. csv-table::
+        :header: "Index", "Role", "Description", "Set by", "Default"
+        :widths: 5, 10, 30, 30, 10
+
+        0, "above", "Above canopy conditions", "``above_ground_canopy_offset``", "+2 m"
+        1, "canopy", "Height of first canopy layer",  "``PlantsModel``", "--"
+        "...", "canopy", "Height of other canopy layers",  "``PlantsModel``", "--"
+        10, "canopy", "Height of the last canopy layer ", "``PlantsModel``", "--"
+        11, "subcanopy", "Subcanopy height", ``subcanopy_layer_height``, "1.5 m"
+        12, "surface", "Near surface conditions", ``surface_layer_height``, "0.1 m"
+        13, "soil", "Upper soil layer depth",  ``soil_layers``, "-0.25 m"
+        14, "soil", "Lower soil layer depth",  ``soil_layers``, "-1.25 m"
+
+    Raises:
+        ConfigurationError: If the configuration elements are incorrect for defining
+            the model timing.
+    """
+
+    canopy_layers: int = field(init=False)
+    """The maximum number of canopy layers."""
+    soil_layers: list[float] = field(init=False)
+    """A list of the depths of soil layer boundaries."""
+    above_canopy_height_offset: float = field(init=False)
+    """The height above the canopy of the provided reference climate variables."""
+    surface_layer_height: float = field(init=False)
+    """The height above ground used to represent surface conditions."""
+    subcanopy_layer_height: float = field(init=False)
+    """The height above ground used to represent subcanopy conditions."""
+    layer_roles: tuple[str, ...] = field(init=False)
+    """An ordered tuple giving the roles of the layers within the model."""
+    config: InitVar[Config]
+    """A validated model configuration."""
+
+    def __post_init__(self, config: Config) -> None:
+        """Populate the ``LayerStructure`` instance.
+
+        This method populates the ``LayerStructure`` attributes from the provided
+        :class:`~virtual_rainforest.core.config.Config` instance.
+
+        Args:
+            config: A Config instance.
+        """
+
+        lyr_config = config["core"]["layers"]
+
+        # Validate contents
+        # NOTE: some of this is also trapped by validation against the core schema.
+        # Canopy layers
+        canopy_layers = lyr_config["canopy_layers"]
+        if (
+            not isinstance(canopy_layers, int)
+            and not (isinstance(canopy_layers, float) and canopy_layers.is_integer())
+            and canopy_layers < 1
+        ):
+            to_raise = ConfigurationError(
+                "The number of canopy layers is not a positive integer."
+            )
+            LOGGER.error(to_raise)
+            raise to_raise
+
+        self.canopy_layers = canopy_layers
```

Comment:
I would make this a separate function, for clarity of the code and easy testing:

```python
def _validate_canopy_layers(canopy_layers) -> int:
    ...

self.canopy_layers = _validate_canopy_layers(lyr_config["canopy_layers"])
```